prober/derp.go

```diff
@@ -282,40 +344,96 @@ func derpProbeNodePair(ctx context.Context, dm *tailcfg.DERPMap, from, to *tailc
 	if err != nil {
+		return err
+	}
+	defer fromc.Close()
+	toc, err := newConn(ctx, dm, to, true)
+	if err != nil {
+		return err
+	}
+	defer toc.Close()
+
+	// Wait a bit for from's node to hear about to existing on the
+	// other node in the region, in the case where the two nodes
+	// are different.
+	if from.Name != to.Name {
+		time.Sleep(100 * time.Millisecond) // pretty arbitrary
+	}
+
+	const meshProbePacketSize = 8
+	if err := runDerpProbeNodePair(ctx, from, to, fromc, toc, meshProbePacketSize); err != nil {
 		// Record pubkeys on failed probes to aid investigation.
-		err = fmt.Errorf("%s -> %s: %w",
+		return fmt.Errorf("%s -> %s: %w",
 			fromc.SelfPublicKey().ShortString(),
 			toc.SelfPublicKey().ShortString(), err)
 	}
-	return latency, err
+	return err
 }
 
-func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client) (latency time.Duration, err error) {
-	// Make a random packet
-	pkt := make([]byte, 8)
-	crand.Read(pkt)
+// probePackets stores a pregenerated slice of probe packets keyed by their total size.
+var probePackets syncs.Map[int64, [][]byte]
 
-	t0 := time.Now()
+// packetsForSize returns a slice of packet payloads with a given total size.
+func packetsForSize(size int64) [][]byte {
+	// For a small payload, create a unique random packet.
+	if size <= derp.MaxPacketSize {
+		pkt := make([]byte, size)
+		crand.Read(pkt)
+		return [][]byte{pkt}
+	}
+
+	// For a large payload, create a bunch of packets once and re-use them
+	// across probes.
+	pkts, _ := probePackets.LoadOrInit(size, func() [][]byte {
+		var pkts [][]byte
+		for remaining := size; remaining > 0; remaining -= derp.MaxPacketSize {
+			pkt := make([]byte, min(remaining, derp.MaxPacketSize))
+			crand.Read(pkt)
+			pkts = append(pkts, pkt)
+		}
+		return pkts
+	})
+	return pkts
+}
 
-	// Send the random packet.
+// runDerpProbeNodePair takes two DERP clients (fromc and toc) connected to two
+// DERP servers (from and to) and sends a test payload of a given size from one
+// to another.
+func runDerpProbeNodePair(ctx context.Context, from, to *tailcfg.DERPNode, fromc, toc *derphttp.Client, size int64) error {
+	pkts := packetsForSize(size)
+
+	// Send the packets.
 	sendc := make(chan error, 1)
 	go func() {
-		sendc <- fromc.Send(toc.SelfPublicKey(), pkt)
+		for _, pkt := range pkts {
+			if err := fromc.Send(toc.SelfPublicKey(), pkt); err != nil {
+				sendc <- err
+				return
+			}
+		}
+		sendc <- nil
 	}()
 	select {
 	case <-ctx.Done():
-		return 0, fmt.Errorf("timeout sending via %q: %w", from.Name, ctx.Err())
+		return fmt.Errorf("timeout sending via %q: %w", from.Name, ctx.Err())
 	case err := <-sendc:
 		if err != nil {
-			return 0, fmt.Errorf("error sending via %q: %w", from.Name, err)
+			return fmt.Errorf("error sending via %q: %w", from.Name, err)
 		}
 	}
 
-	// Receive the random packet.
-	recvc := make(chan any, 1) // either derp.ReceivedPacket or error
+	// Receive the packets.
```

Comment:
If there is more than one, won't we have to make sure to drain multiple responses below?

Reply:
I think we've only ever returned a single error, but I've restructured this following your recommendation.